File: 5. DetectCycleDirected.cpp
/*
Task: Detect Cycle in Directed Graph
Write a function to determine if a given directed graph contains a cycle.
*/
#include <iostream>
#include <vector>
using namespace std;

bool hasCycleUtil(int u, vector<vector<int>>& adj, vector<bool>& visited, vector<bool>& recStack)
{
    visited[u] = recStack[u] = true;
    for (int v : adj[u])
    {
        if (!visited[v] && hasCycleUtil(v, adj, visited, recStack))
            return true;
        else if (recStack[v])
            return true;
    }
    recStack[u] = false;
    return false;
}
bool hasCycleDirected(int V, vector<vector<int>>& adj)
{
    vector<bool> visited(V, false), recStack(V, false);
    for (int u = 0; u < V; ++u)
    {
        if (!visited[u])
        {
            if (hasCycleUtil(u, adj, visited, recStack))
                return true;
        }
    }
    return false;
}
// Sample Test Case
int main()
{
    int V = 4;
    vector<vector<int>> adj(V);
    adj[0] = {1};
    adj[1] = {2};
    adj[2] = {3};
    adj[3] = {1};  // cycle
    cout << "Cycle (Directed): " << (hasCycleDirected(V, adj) ? "Yes" : "No") << endl;
    return 0;
}
/*
Time Complexity: O(V+E)
Space Complexity: O(V)
*/

File: 5. DetectCycleDirected.py
"""
Task: Detect Cycle in Directed Graph
Write a function to determine if a given directed graph contains a cycle.
"""
def has_cycle_util(u, adj, visited, recStack):
    visited[u] = recStack[u] = True
    for v in adj[u]:
        if not visited[v]:
            if has_cycle_util(v, adj, visited, recStack):
                return True
        elif recStack[v]:
            return True
    recStack[u] = False
    return False
def has_cycle_directed(V, adj):
    visited = [False]*V
    recStack = [False]*V
    for u in range(V):
        if not visited[u]:
            if has_cycle_util(u, adj, visited, recStack):
                return True
    return False
# Sample Test Case
if __name__ == "__main__":
    V = 4
    adj = [[1], [2], [3], [1]] # cycle
    print("Cycle (Directed):", has_cycle_directed(V, adj))
"""
Time Complexity: O(V+E)
Space Complexity: O(V)
"""
